Add infer meta

phlrain opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
Breaking changes

### PR changes
Others

### Describe
<!-- Describe what this PR does -->
add norm，batchnorm dropout， conv infer meta

phlrain committed:
add infer meta; test=develop

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

phlrain committed:
add histogram infer meta; test=develop

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

phlrain committed:
fix unitest bug; test=develop

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

phlrain committed:
format; test=develop

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

phlrain committed:
format; test=develop

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

phlrain committed:
bn not use new infer meta; test=develop

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

chenwhql commented on paddle/fluid/operators/batch_norm_op.cc:
```diff
@@ -951,335 +521,16 @@ framework::OpKernelType BatchNormDoubleGradOp::GetExpectedKernelType(
+DECLARE_INFER_SHAPE_FUNCTOR(batch_norm, BatchNormInferShapeFunctor,
+                            PD_INFER_META(phi::BatchNormInferMeta));
+
 REGISTER_OPERATOR(batch_norm, ops::BatchNormOp, ops::BatchNormOpMaker,
```
这里BatchNormInferShapeFunctor要加到REGISTER_OPERATOR中

chenwhql commented on paddle/fluid/operators/conv_op.cc:
```diff
@@ -841,6 +844,8 @@ framework::OpKernelType ConvOpDoubleGrad::GetExpectedKernelType(
 namespace ops = paddle::operators;
+DECLARE_INFER_SHAPE_FUNCTOR(conv2d, Conv2dInferShapeFunctor,
+                            PD_INFER_META(phi::ConvInferMeta));
 REGISTER_OPERATOR(conv2d, ops::ConvOp, ops::Conv2DOpMaker,
```
同上，有多处，可以一并检查下

chenwhql commented on paddle/phi/infermeta/multiary.cc:
```diff
@@ -200,6 +201,114 @@ void AucInferMeta(const MetaTensor& input,
+  mean_out->set_dims({C});
+  variance_out->set_dims({C});
+  saved_mean->set_dims({C});
+  saved_variance->set_dims({C});
```
同时set_dtype一下？

chenwhql commented on paddle/phi/infermeta/unary.cc:
```diff
@@ -392,6 +403,26 @@ void GumbelSoftmaxInferMeta(const MetaTensor& x,
+                                   max,
+                                   min));
+
+  out->set_dims({bins});
```
set_dtype

chenwhql commented on python/paddle/fluid/tests/unittests/test_optimizer.py:
```diff
@@ -29,931 +29,920 @@
-                                       "dropout_with_seed_cpu.tmp_1.subprog_0"
-                                   ])
-                self.assertEqual(drop_vec[0].tolist(), drop_vec[1].tolist())
+# class TestOptimizer(unittest.TestCase):
```
这里在合入前需要打开吧

chenwhql reviewed (commented):
conv和batch norm需要顺便增加一下yaml，或者先在对应infermeta文件下注册infermeta映射关系，否则infrt还是不能用

phlrain committed:
add infer meta; test=develop

phlrain committed:
fixbug; test=develop

phlrain committed:
fix bug; test=develofor (const auto& iter : grad_attrs) {
    if (IgnoreGradAttribute(fwd_op_type, iter.first)) continue;
    std::pair<std::string, std::string> type_val =
        GetAttrType(iter.second, false /*is_arg*/);
    const char* GRAD_ATTRS_TEMPLATE =
        "  %s %s = %s;\n"
        "  %s[\"%s\"] = %s;\n";
    std::string var_name = iter.first + std::to_string(*outs_size);
    grad_attrs_str += paddle::string::Sprintf(
        GRAD_ATTRS_TEMPLATE, type_val.first, var_name, type_val.second,
        attrs_name, iter.first, var_name);
  }

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

phlrain committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into add_infer_meta

phlrain committed:
recover unitest; test=develop

chenwhql reviewed (approved):
细节后续完善
